### PR TITLE
[callhierarchy] added call hierarchy for Go

### PR DESCRIPTION
**Pull request by JanKoehnlein**
Signed-off-by: Jan Koehnlein <jan.koehnlein@typefox.io>

### CODE REVIEW

**Comment by JanKoehnlein** [comment]
Available in the context menu over a function or function call. 
Please test whether it makes sense this way.

**Comment by simark** [comment]
If I understand correctly, the call hierarchy UI is not really decided by this extension, it's done by some other Theia extension or the core?  So what this extension does is simply state that the Go language supports call hierarchies?

I used this code for testing: https://github.com/simark/ls-interact/tree/master/go-test

When I try on the `plus` and `plusPlus` functions, it works.  When I try on the `Times` function (which is defined in other.go), I get:

```
Error running guru: Command failed: /home/emaisin/gopath/bin/guru -modified referrers /home/emaisin/src/ls-interact/go-test/other.go:#19
guru: query object not found during reloading
```

in an error popup.

And when trying on Println:

```
Command failed: /home/emaisin/gopath/bin/godef -t -i -f /home/emaisin/src/ls-interact/go-test/test.go -o 830
godef: no declaration found for fmt.Println
```

**Comment by JanKoehnlein** [comment]
@simark Yes, the callhierarchy is implemented in `@theia/callhierarchy`. We just state here that we support it.

The problem you describe comes from a missconfigured GOPATH variable. The command line tools the go LS delegates to needs GOPATH to be defined correctly to find references across files. By default, it assumes `GOPATH=~/go`. So this is not a callhierarchy issue, but a general one.

I've created https://github.com/theia-ide/theia-go-extension/issues/12 to discuss this.

A workaround here is to put 
```json
{
    "inferGopath": true
}
```
in the `go.json` in the root of your workspace.

File: go/src/browser/go-calhierarchy-service.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
```
Comment (simark) [inline go/src/browser/go-calhierarchy-service.ts]: 2018?

File: go/src/browser/go-frontend-module.ts [hunk 2]
```diff
@@ -13,11 +13,15 @@ import { MenuContribution, CommandContribution } from '@theia/core';
...
 import { GoMenuContribution } from './go-menu-contribution';
+import { GoCallHierarchyService } from "./go-calhierarchy-service";
+import { CallHierarchyService } from "@theia/callhierarchy/lib/browser/callhierarchy-service";
 
 export default new ContainerModule((bind: interfaces.Bind, unbind: interfaces.Unbind, isBound: interfaces.IsBound, rebind: interfaces.Rebind) => {
     bind(LanguageClientContribution).to(GoClientContribution).inSingletonScope();
     bind(MenuContribution).to(GoMenuContribution).inSingletonScope();
     bind(CommandContribution).to(GoCommandContribution).inSingletonScope();
 	bind(GoCommands).to(GoCommands).inSingletonScope();
+	bind(GoCallHierarchyService).toSelf().inSingletonScope();
```
Comment (simark) [inline go/src/browser/go-frontend-module.ts:24]: There is some indentation issue here.

**Review by simark — APPROVED** [review]

**Pull request merged**